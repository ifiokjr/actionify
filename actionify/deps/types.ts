export * from "https://cdn.skypack.dev/ts-brand?dts";
export * from "https://cdn.skypack.dev/type-fest@2.18.0?dts";
