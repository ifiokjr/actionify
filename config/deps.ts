export {
  objectEntries,
  objectKeys,
} from "https://cdn.skypack.dev/ts-extras@0.11.0?dts";
export { parse } from "https://deno.land/std@0.156.0/flags/mod.ts";
export * from "https://deno.land/std@0.156.0/fs/mod.ts";
export * from "https://deno.land/std@0.156.0/fs/mod.ts";
export { readLines } from "https://deno.land/std@0.156.0/io/mod.ts";
export { getLogger } from "https://deno.land/std@0.156.0/log/mod.ts";
export * as semver from "https://deno.land/std@0.156.0/semver/mod.ts";
export { globber } from "https://deno.land/x/globber@0.1.0/mod.ts";
export {
  html,
  type Token,
  tokens,
} from "https://deno.land/x/rusty_markdown@v0.4.1/mod.ts";
export { udd, type UddResult } from "https://deno.land/x/udd@0.7.5/mod.ts";
export { default as gitUrlParse } from "https://esm.sh/git-url-parse@12.0.0";
export { default as pLimit } from "https://esm.sh/p-limit@4.0.0";
