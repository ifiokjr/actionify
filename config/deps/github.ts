export * as coreActions from "https://esm.sh/@actions/core@1.9.0";
export * as githubActions from "https://esm.sh/@actions/github@5.0.3";
