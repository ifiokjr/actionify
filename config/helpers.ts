import { gitUrlParse, readLines, semver } from "./deps.ts";

function clean(version: string) {
  return semver.parse(version.trim().replace(/^[=v]+/, ""))?.version;
}

export async function getTagVersion(cwd = Deno.cwd()): Promise<string> {
  let version = "0.0.0";
  const stdout = Deno.run({ cmd: ["git", "tag"], cwd, stdout: "piped" }).stdout;

  try {
    for await (const line of readLines(stdout)) {
      const cleaned = clean(line) ?? version;
      version = semver.gt(cleaned, version) ? cleaned : version;
    }
  } catch {
    // Do nothing
  }

  return version;
}

/**
 * Get the currently checked out commit sha.
 */
export async function getCurrentCommit(cwd = Deno.cwd()): Promise<string> {
  const stdout = await Deno.run({
    cmd: ["git", "rev-parse", "HEAD"],
    cwd,
    stdout: "piped",
  }).output();

  return new TextDecoder().decode(stdout).trim();
}

export async function gitCheckoutOrCreate(name: string) {
  // try to checkout out
  const result = await Deno.run({
    cmd: ["git", "checkout", name],
    cwd: cwd.pathname,
  }).status();

  if (result.success) {
    return;
  }

  await Deno.run({
    cmd: ["git", "checkout", "-b", name],
    cwd: cwd.pathname,
  }).status();
}

/**
 * Read the remote URL.
 */
export async function getGitHubRemote(cwd = Deno.cwd()): Promise<GitHubRemote> {
  const stdout = await Deno.run({
    cmd: ["git", "remote", "get-url", "origin"],
    cwd,
    stdout: "piped",
  }).output();
  const remote = new TextDecoder().decode(stdout);
  const parsed = gitUrlParse(remote);

  return {
    fullName: parsed.full_name,
    href: `https://${parsed.source}/${parsed.full_name}/`,
    name: parsed.name,
    owner: parsed.owner,
  };
}

interface GitHubRemote {
  owner: string;
  name: string;
  fullName: string;
  href: string;
}

/**
 * Read the first commit in the commit history.
 */
export async function getFirstCommit(cwd = Deno.cwd()) {
  const stdout = await Deno.run({
    cmd: ["git", "rev-list", "--max-parents=0", "HEAD"],
    cwd,
    stdout: "piped",
  }).output();
  const ref = new TextDecoder().decode(stdout);

  return ref.slice(0, 7);
}

/**
 * Create a tag for the provided version.
 */
export function createTag(
  version: string,
  cwd = Deno.cwd(),
): Promise<Deno.ProcessStatus> {
  if (!semver.valid(version)) {
    throw new Error(`Invalid tag version provided: ${version}`);
  }

  // NOTE: it's important we use the -m flag to create annotated tag otherwise
  // 'git push --follow-tags' won't actually push the tags
  return Deno
    .run({ cmd: ["git", "tag", "version", "-m", version], cwd })
    .status();
}

export async function gitReset(
  ref: string,
  mode: "hard" | "soft" | "mixed" = "hard",
) {
  await Deno.run({
    cmd: ["git", "reset", `--${mode}`, ref],
    cwd: cwd.pathname,
    stdout: "inherit",
  }).status();
}

export const cwd = new URL("..", import.meta.url);
