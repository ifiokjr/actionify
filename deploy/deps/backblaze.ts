export { default as BackBlaze } from "https://esm.sh/backblaze-b2@1.7.0";
