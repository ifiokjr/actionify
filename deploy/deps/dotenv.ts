import "https://deno.land/std@0.155.0/dotenv/load.ts";
