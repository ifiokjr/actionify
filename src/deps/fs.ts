export * from "https://deno.land/std@0.155.0/fs/mod.ts";
