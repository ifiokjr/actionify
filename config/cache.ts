import { getLogger, globber, parse } from "./deps.ts";
import { cwd } from "./helpers.ts";

const decoder = new TextDecoder();
const log = getLogger();
const deps: string[] = [];
const args = parse(Deno.args, { boolean: ["reload"] });
const entries = globber({
  cwd,
  include: ["**/deps.ts", "**/deps/*.ts", "import_map.json"],
  excludeDirectories: true,
});

for await (const entry of entries) {
  deps.push(entry.relative);
}

const baseCommand = [
  "deno",
  "cache",
  "--lock=lock.json",
  "--import-map=import_map.json",
];

async function update() {
  log.info("Updating the `lock.json` file.");

  await Deno.run({
    cmd: [...baseCommand, "--reload", "--lock-write", ...deps],
    stdout: "piped",
    cwd: cwd.pathname,
  }).output();
}

async function load() {
  const CI = !!Deno.env.get("CI");
  log.info("Loading the cache from `lock.json`.");

  const command = Deno.run({
    cmd: [...baseCommand, ...deps],
    stdout: "inherit",
    stderr: "piped",
    cwd: cwd.pathname,
  });

  const [status, stderr] = await Promise.all([
    command.status(),
    command.stderrOutput(),
  ]);
  command.close();

  if (!status.success) {
    const error = decoder.decode(stderr);

    if (error.includes("No such file or directory") && !CI) {
      log.warning("No `lock.json` found. Creating a new one.");
      await update();
    } else {
      log.info(error);
      log.critical(
        `Error while reloading the cache.\n${error.split("error: ")[1]}`,
      );
      Deno.exit(1);
    }
  }
}

if (args.reload) {
  await load();
} else {
  await update();
}
