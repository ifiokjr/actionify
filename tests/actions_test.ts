import { generateTypeScriptFromAction } from "https://deno.land/x/actionify@0.1.0/mod.ts";
import { emptyDir, ensureDir } from "../src/deps/fs.ts";
import { assert } from "./deps.ts";
import { snapshot } from "./helpers.ts";

Deno.test("generate TypeScript module from cache.yml", async (t) => {
  const result = await generateTypeScriptFromAction(
    import.meta.resolve("./fixtures/actions/cache.yml"),
    "actions/cache@3.0.8",
  );

  await snapshot(t, result);
  await assertCanRun(
    result,
    { path: "this/is/awesome", key: "this is a key" },
    {},
    "cache",
  );
});

Deno.test("generate TypeScript module from checkout.yml", async (t) => {
  const result = await generateTypeScriptFromAction(
    import.meta.resolve("./fixtures/actions/checkout.yml"),
    "actions/checkout@3.0.2",
  );

  await snapshot(t, result);

  await assertCanRun(result, { ref: "main" }, { reff: "" }, "checkout");
});

async function assertCanRun(
  content: string,
  props: object,
  failingProps: object,
  name: string,
) {
  const pathToMod = import.meta.resolve("../mod.ts");
  const action = `\
import testStep from "./test_step.ts";
import { workflow, defineWorkflows } from "${pathToMod}";

const test = workflow({ name: "test" })
  .on("push")
  .job("awesome", (job) => {
    return job
      .runsOn("ubuntu-latest")
      .step(testStep(${JSON.stringify(props)}))
      // @ts-expect-error for the purpose of the test.
      .step(testStep(${JSON.stringify(failingProps)}));
  });

export default defineWorkflows({
  workflows: [test],
  rootDirectory: import.meta.resolve("./"),
});
`;
  const root = new URL(import.meta.resolve(`./fixtures/tmp/${name}/`));
  const cliPath = new URL(import.meta.resolve("../cli.ts"));

  await ensureDir(root);
  const contentPath = new URL("./test_step.ts", root);
  const actionPath = new URL("./actionify.ts", root);

  await Deno.writeTextFile(contentPath, content);
  await Deno.writeTextFile(actionPath, action);

  const checkProcess = Deno.run({
    cwd: root.pathname,
    cmd: ["deno", "check", "-r", actionPath.pathname],
    stderr: "null",
    stdout: "null",
  });
  const checkStatus = await checkProcess.status();
  checkProcess.close();

  assert(checkStatus.success, "should successfully typecheck");

  const buildProcess = Deno.run({
    cwd: root.pathname,
    stderr: "null",
    stdout: "null",
    cmd: [
      "deno",
      "run",
      "-Ar",
      cliPath.pathname,
      "-c",
      actionPath.pathname,
      "-o",
      root.pathname,
    ],
  });
  const buildStatus = await buildProcess.status();
  buildProcess.close();

  assert(buildStatus.success, "should be able to use the custom module");
  await emptyDir(root);
}
