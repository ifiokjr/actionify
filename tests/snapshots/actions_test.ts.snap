export const snapshot = {};

snapshot[`generate TypeScript module from cache.yml 1`] = `
import {
  type ActionTemplate,
  type DefaultStepInputs,
  e,
  step,
  type WithContext,
} from "https://deno.land/x/actionify@0.1.0/mod.ts";

export interface CacheInputs extends DefaultStepInputs {
  /** A list of files, directories, and wildcard patterns to cache and restore */
  "path": e.ExpressionValue;

  /** An explicit key for restoring and saving the cache */
  "key": e.ExpressionValue;

  /** An ordered list of keys to use for restoring stale cache if no cache hit occurred for key. Note \`cache-hit\` returns false in this case. */
  "restore-keys"?: e.ExpressionValue;

  /** The chunk size used to split up large files during upload, in bytes */
  "upload-chunk-size"?: e.ExpressionValue;
}

/** Cache artifacts like dependencies and build outputs to improve workflow execution time */
export default function cache<Base extends ActionTemplate = ActionTemplate>(inputs: WithContext<CacheInputs, Base, "jobs:jobId:steps:with">) {
  return step<Base & { stepOutputs: "cache-hit" }>()
    .uses("actions/cache@3.0.8")
    // @ts-expect-error Fix this later
    .with(inputs as DefaultStepInputs);
}

export const NAME = "Cache";
export const AUTHOR = "GitHub";

`;

snapshot[`generate TypeScript module from checkout.yml 1`] = `
import {
  type ActionTemplate,
  type DefaultStepInputs,
  e,
  step,
  type WithContext,
} from "https://deno.land/x/actionify@0.1.0/mod.ts";

export interface CheckoutInputs extends DefaultStepInputs {
  /**
   * Repository name with owner. For example, actions/checkout
   * @default "\${{ github.repository }}"
   */
  "repository"?: e.ExpressionValue;

  /**
   * The branch, tag or SHA to checkout. When checking out the repository that triggered a workflow, this defaults to the reference or SHA for that event.  Otherwise, uses the default branch.
   * 
   */
  "ref"?: e.ExpressionValue;

  /**
   * Personal access token (PAT) used to fetch the repository. The PAT is configured with the local git config, which enables your scripts to run authenticated git commands. The post-job step removes the PAT.
   * 
   * We recommend using a service account with the least permissions necessary. Also when generating a new PAT, select the least scopes necessary.
   * 
   * [Learn more about creating and using encrypted secrets](https://help.github.com/en/actions/automating-your-workflow-with-github-actions/creating-and-using-encrypted-secrets)
   * 
   * @default "\${{ github.token }}"
   */
  "token"?: e.ExpressionValue;

  /**
   * SSH key used to fetch the repository. The SSH key is configured with the local git config, which enables your scripts to run authenticated git commands. The post-job step removes the SSH key.
   * 
   * We recommend using a service account with the least permissions necessary.
   * 
   * [Learn more about creating and using encrypted secrets](https://help.github.com/en/actions/automating-your-workflow-with-github-actions/creating-and-using-encrypted-secrets)
   * 
   */
  "ssh-key"?: e.ExpressionValue;

  /**
   * Known hosts in addition to the user and global host key database. The public SSH keys for a host may be obtained using the utility \`ssh-keyscan\`. For example, \`ssh-keyscan github.com\`. The public key for github.com is always implicitly added.
   * 
   */
  "ssh-known-hosts"?: e.ExpressionValue;

  /**
   * Whether to perform strict host key checking. When true, adds the options \`StrictHostKeyChecking=yes\` and \`CheckHostIP=no\` to the SSH command line. Use the input \`ssh-known-hosts\` to configure additional hosts.
   * 
   * @default true
   */
  "ssh-strict"?: e.ExpressionValue;

  /**
   * Whether to configure the token or SSH key with the local git config
   * @default true
   */
  "persist-credentials"?: e.ExpressionValue;

  /** Relative path under \$GITHUB_WORKSPACE to place the repository */
  "path"?: e.ExpressionValue;

  /**
   * Whether to execute \`git clean -ffdx && git reset --hard HEAD\` before fetching
   * @default true
   */
  "clean"?: e.ExpressionValue;

  /**
   * Number of commits to fetch. 0 indicates all history for all branches and tags.
   * @default 1
   */
  "fetch-depth"?: e.ExpressionValue;

  /** Whether to download Git-LFS files */
  "lfs"?: e.ExpressionValue;

  /**
   * Whether to checkout submodules: \`true\` to checkout submodules or \`recursive\` to recursively checkout submodules.
   * 
   * When the \`ssh-key\` input is not provided, SSH URLs beginning with \`git@github.com:\` are converted to HTTPS.
   * 
   */
  "submodules"?: e.ExpressionValue;

  /**
   * Add repository path as safe.directory for Git global config by running \`git config --global --add safe.directory <path>\`
   * @default true
   */
  "set-safe-directory"?: e.ExpressionValue;
}

/** Checkout a Git repository at a particular version */
export default function checkout<Base extends ActionTemplate = ActionTemplate>(inputs: WithContext<CheckoutInputs, Base, "jobs:jobId:steps:with"> = {}) {
  return step<Base>()
    .uses("actions/checkout@3.0.2")
    // @ts-expect-error Fix this later
    .with(inputs as DefaultStepInputs);
}

export const NAME = "Checkout";
export const AUTHOR = "";

`;
